### PR TITLE
Add thread safety, robustness, and filesystem tests

### PR DESCRIPTION
**Pull request by marcuschangarm**
* Use mutex to ensure each operation is atomic.
 * Add explicit software reset during initialization.
 * Add filesystem tests.
 * Updated documentation.

### CODE REVIEW

**Comment by marcuschangarm** [comment]
I didn't use the SPI mutex after all since that would lock the SPI bus longer than necessary. 

The not-write-protect pin comes after the frequency so that the arguments are compatible with the spif-driver.

The software reset is there because the device seem to enter an undefined state sometimes.

File: DataFlashBlockDevice.cpp [hunk 1]
```diff
@@ -127,6 +128,8 @@ enum dummy {
     DATAFLASH_HIGHEST_FREQUENCY_BYTES = 2
 };
 
+SingletonPtr<PlatformMutex> DataFlashBlockDevice::_mutex;
```
Comment (geky) [inline DataFlashBlockDevice.cpp:131]: Why singleton pointer? To protect the SPI bus? Should you use the actual SPI locking?

Comment (geky) [reply DataFlashBlockDevice.cpp:131]: Ah just saw the above comment. Shouldn't you then use a member mutex combined with the SPI locking? Unfortunately this won't synchronize with another device on the same SPI bus

Comment (marcuschangarm) [reply DataFlashBlockDevice.cpp:131]: Ah, so you are saying every time we lock the dataflash mutex, we also lock the SPI mutex for the entire duration?

Comment (geky) [reply DataFlashBlockDevice.cpp:131]: Well, you only need to lock the SPI mutex for as long as you need the bus uninterrupted. But if there's additional state you need to protect outside of the SPI bus you can use an additional mutex that wraps the logic and bus.

If the logic modification quick, you could just use the SPI bus's lock for all synchronization, but that's up to you.

**Pull request closed without merging**